File: sdk/src/utils/crypto.ts
import * as jose from 'jose';
import type { SigningOptions } from '../types.js';

/**
 * Generate an Ed25519 or ES256 key pair for signing and verifying tokens.
 */
export async function generateKeyPair(
  algorithm: 'EdDSA' | 'ES256' = 'EdDSA'
): Promise<CryptoKeyPair> {
  const alg = algorithm === 'EdDSA' ? 'EdDSA' : 'ES256';
  const { publicKey, privateKey } = await jose.generateKeyPair(alg, {
    extractable: true,
  });
  return { publicKey, privateKey } as CryptoKeyPair;
}

/**
 * Sign a payload, producing a compact JWS string.
 */
export async function sign(
  payload: Record<string, unknown>,
  options: SigningOptions
): Promise<string> {
  const alg = options.algorithm === 'ES256' ? 'ES256' : 'EdDSA';
  const header: Record<string, unknown> = { alg };
  if (options.kid) {
    header.kid = options.kid;
  }

  const jws = await new jose.CompactSign(
    new TextEncoder().encode(JSON.stringify(payload))
  )
    .setProtectedHeader(header as jose.CompactJWSHeaderParameters)
    .sign(options.privateKey);

  return jws;
}

/**
 * Verify a compact JWS string and return the decoded payload.
 */
export async function verify(
  token: string,
  key: CryptoKey
): Promise<Record<string, unknown>> {
  const { payload } = await jose.compactVerify(token, key);
  return JSON.parse(new TextDecoder().decode(payload));
}

File: sdks/python/apoa/crypto.py
"""Key generation, JWT signing, and verification. Ed25519 (EdDSA) and ES256."""

from __future__ import annotations

import json
from base64 import urlsafe_b64decode, urlsafe_b64encode
from typing import Any

import jwt
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.asymmetric.ec import (
    SECP256R1,
    EllipticCurvePrivateKey,
    EllipticCurvePublicKey,
    generate_private_key as generate_ec_private_key,
)

from .types import SigningOptions


def generate_key_pair(
    algorithm: str = "EdDSA",
) -> tuple[Any, Any]:
    """Generate a key pair for token signing.

    Returns (private_key, public_key).
    """
    if algorithm == "ES256":
        private_key = generate_ec_private_key(SECP256R1())
        return private_key, private_key.public_key()

    # Default: EdDSA (Ed25519)
    private_key = Ed25519PrivateKey.generate()
    return private_key, private_key.public_key()


def sign_token(payload: dict[str, Any], options: SigningOptions) -> str:
    """Sign a payload as a compact JWS (JWT)."""
    alg = "ES256" if options.algorithm == "ES256" else "EdDSA"
    headers: dict[str, str] = {"alg": alg}
    if options.kid:
        headers["kid"] = options.kid

    return jwt.encode(payload, options.private_key, algorithm=alg, headers=headers)


def verify_token(raw_jwt: str, public_key: Any) -> dict[str, Any]:
    """Verify a JWT signature and return the decoded payload."""
    # PyJWT needs to know which algorithms to accept
    return jwt.decode(
        raw_jwt,
        public_key,
        algorithms=["EdDSA", "ES256"],
        options={"verify_exp": False, "verify_aud": False},
    )


def decode_header(raw_jwt: str) -> dict[str, Any]:
    """Decode a JWT header without verifying the signature."""
    return jwt.get_unverified_header(raw_jwt)
